### PR TITLE
[Blog] React components with multiple slots

### LINKED ISSUE #518: [Blog] Composants React avancé avec slots
Comment créer des composants React avec des slots en tant qu'enfants JSX clairement identifiés. 
Alternative plus lisible par rapport à utiliser des props pour ce genre de besoins.

Ex: 

- `DropMenu` & items : https://github.com/Elao/onisep/blob/8ea28ba897ed05865ac43b9933a41769ac66d11c/front/src/pages/Film/Header/HeaderNavItems.tsx#LL92C25-L111C18:

    ```jsx
      <DropMenu>
        <DropMenu.Trigger>
          <button type="button">
            {capitalize((profile as UserProfile).username)}
            <span className="fr-icon-arrow-down-s-line fr-fi--sm"></span>
          </button>
        </DropMenu.Trigger>

        <DropMenu.Items>
          <DropMenu.Item>
            <Link to={route(ListTemplates)}>Espace administrateur</Link>
          </DropMenu.Item>
          <DropMenu.Item>
            <Link to={route(ListFilms)}>Espace enseignant</Link>
          </DropMenu.Item>
          <DropMenu.Item>
            <Link to={route(Logout)}>Déconnexion</Link>
          </DropMenu.Item>
        </DropMenu.Items>
      </DropMenu>
    ```

- Modal and header, footer, content, actions, … https://github.com/Elao/onisep/blob/8ea28ba897ed05865ac43b9933a41769ac66d11c/front/src/pages/Film/Share/PrivateLinkModal.tsx#L30-L58 : 
```jsx
<Modal hide={close} isOpen={opened}>
    <Modal.Title>
      Le lien de votre film est unique, pensez à le sauvegarder !
    </Modal.Title>

    <Modal.Content>
      <label className="fr-label fr-sr-only" htmlFor="share-link">Copier le lien</label>

      <CopyUrlInput id="share-link" url={privateLink} copyLabel="Copier" />

      <p>
        Le lien généré pour votre film est <strong>unique et irremplaçable</strong>,
        si vous l&apos;égarez tout votre travail sera perdu !
      </p>
      <p>Gardez précieusement ce lien pour retrouver votre film et l&apos;ensemble des modifications réalisées.</p>
    </Modal.Content>

    <Modal.Footer>
      <div className="fr-btns-group fr-btns-group--inline fr-btns-group--right fr-btns-group--icon-left">
        <button className="fr-btn" onClick={onConfirm}>
            J&apos;ai compris
        </button>
      </div>
    </Modal.Footer>
</Modal>
```

### PR DESCRIPTION
**Pull request by ogizanagi**
Preview: https://elao.github.io/elao_/pr/519/blog/dev/react-component-slots/

Fixes #518 

### TODO

- [x] Schema to represent the modal slots? Which tool to use to generate a cool visualization? ➜ https://excalidraw.com/ (thx @Romaneled )

### CODE REVIEW

File: content/blog/dev/react-component-slots.md [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+thumbnail:          "content/images/blog/thumbnails/react-component-slots.jpg"
+tags:               ["React", "Slots"]
+authors:            ["msteinhausser"]
+credits:            { name: "Lautaro Andreani", url: "https://unsplash.com/@lautaroandreani" }
+tweetId:            ~
+---
+
+## Introduction
+
+En tant que développeur React, vous avez probablement déjà été confronté à la création de composants génériques.
```
Comment (xavier-rdo) [inline content/blog/dev/react-component-slots.md:21]: [nitpick] 

```diff
- confronté
+ confrontés
```

File: content/blog/dev/react-component-slots.md [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+    </div>
+
+    <div class="modal-footer">
+        <!-- Footer -->
+        <button onClick="…">Fermer</button>
+    </div>
+</div>
+```
+
+Chacune des sous-parties `Title`, `Content` et `Footer` de ce composant possède autour de leur contenu un markup HTML
```
Comment (xavier-rdo) [inline content/blog/dev/react-component-slots.md:54]: [nitpick] `chacune` est par essence au singulier ...
```diff
- de leur contenu
+ de son contenu
```

File: content/blog/dev/react-component-slots.md [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+    </div>
+    
+    <div className="modal-footer">
+      {footer({ close })}
+    </div>
+  </div>;
+}
+```
+
+**Inconvénient** : à fur et à mesure que vos composants, le nombre de slots, ou que les contenus injectés
```
Comment (xavier-rdo) [inline content/blog/dev/react-component-slots.md:229]: [nitpick] 

```diff
- à fur et à mesure que
+ au fur et à mesure que
Encore mieux : 
+ à mesure que
```

```diff
- ou que les contenus grossissent
+ ou les contenus augmentent
```

File: content/blog/dev/react-component-slots.md [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+  </>}
+  footer={({ close }) => <>
+    <button onClick={{/* … */}}>Annuler</button>
+    <button onClick={close}>Accepter</button>
+  </>}
+/>
+```
+
+L'intérêt du [JSX](https://fr.reactjs.org/docs/introducing-jsx.html) étant d'exposer une syntaxe proche du HTML, il est
+moins naturel à la lecture de retrouver ces contenus en tant que propriétés plutôt que nœuds enfant d'un composant.
```
Comment (xavier-rdo) [inline content/blog/dev/react-component-slots.md:257]: [nitpick] 

```diff
- noeuds enfant
+ noeuds enfants
```

File: content/blog/dev/react-component-slots.md [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+
+Vous pouvez consulter sur [ce gist](https://gist.github.com/ogizanagi/db03004d72b680ab03998308f3c18823) 
+l'implémentation d'un composant `<DropMenu>` utilisant cette approche, TypeScript et 
+[MUI Base](https://mui.com/base/getting-started/overview/), relatif à l'exemple de notre [introduction](#abstraire-les-comportements).
+
+## Conclusion
+
+React ne suggère pas de manière explicite la notion de slots, mais il est possible de s'inspirer de ce concept en
+exploitant les propriétés de JSX et `children` afin de simuler l'utilisation de slots multiples avec une syntaxe proche
+du HTML, semblable à ce qu'il peut se faire avec d'autres frameworks.
```
Comment (xavier-rdo) [inline content/blog/dev/react-component-slots.md:394]: [nitpick]

```diff
- à ce qu'il peut se faire
+ à ce qui peut se faire
```

**Review by xavier-rdo — APPROVED** [review]
Super article limpide, pédago et utile !!! Merci Maxime ! Vraiment relevé des broutilles ...

**Comment by ogizanagi** [comment]
> Super article limpide, pédago et utile !!! Merci Maxime ! Vraiment relevé des broutilles ...

Merci 🙏🏻

**Pull request merged**